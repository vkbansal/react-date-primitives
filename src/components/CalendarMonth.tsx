import * as React from 'react';

import {
    startOfMonth,
    isDayAfter,
    isDayBefore,
    isSameDay,
    setMonth,
    getDaysOfMonth
} from '../utils';

import { CalendarMonthProps, DayOfMonth, CalendarDropdownOption } from './types';

/**
 * Primitive react component that can be used to make a datepicker.
 */
export class CalendarMonth extends React.Component<CalendarMonthProps> {
    private getDaysofMonth(props: CalendarMonthProps): DayOfMonth[][] {
        const { month, startDate, endDate, minDate, maxDate } = props;
        const currentMonth = getDaysOfMonth(month);

        const startDateDisabled = !!startDate && !!minDate && !isDayAfter(startDate, minDate);

        return currentMonth.map(
            (week): DayOfMonth[] =>
                week.map(
                    (day): DayOfMonth => {
                        const { date, inCurrentMonth } = day;
                        const disabled =
                            !inCurrentMonth ||
                            Boolean(
                                (minDate && isDayBefore(date, minDate)) ||
                                    (maxDate && isDayAfter(date, maxDate))
                            );

                        const selected =
                            !disabled &&
                            Boolean(
                                (startDate && isSameDay(date, startDate)) ||
                                    (endDate && isSameDay(date, endDate))
                            );

                        const inRange = Boolean(
                            startDate &&
                                isDayAfter(date, startDate) &&
                                endDate &&
                                isDayBefore(date, endDate)
                        );

                        return {
                            date,
                            inRange:
                                inCurrentMonth &&
                                !startDateDisabled &&
                                !disabled &&
                                (selected || inRange),
                            selected,
                            disabled,
                            inCurrentMonth
                        };
                    }
                )
        );
    }

    private getDropdowns = (
        props: CalendarMonthProps
    ): [CalendarDropdownOption[], CalendarDropdownOption[]] => {
        const { minDate, maxDate, month } = props;

        const firstDay = startOfMonth(month);
        const today = new Date();
        const minYear = minDate ? minDate.getFullYear() : today.getFullYear() - 50;
        const maxYear = maxDate ? maxDate.getFullYear() + 1 : today.getFullYear() + 5;

        const dropdownMonths: CalendarDropdownOption[] = Array.from({ length: 12 }, (_, i) => ({
            disabled: Boolean(
                (minDate && isDayBefore(setMonth(month, i), minDate)) ||
                    (maxDate && isDayAfter(setMonth(month, i), maxDate))
            ),
            value: i,
            selected: firstDay.getMonth() === i
        }));

        const dropdownYears: CalendarDropdownOption[] = Array.from(
            { length: maxYear - minYear },
            (_, i) => ({
                value: i + minYear,
                disabled: false,
                selected: i + minYear === firstDay.getFullYear()
            })
        );

        return [dropdownMonths, dropdownYears];
    };

    render() {
        const { month, render, showDropdowns } = this.props;

        let monthsDropdown: CalendarDropdownOption[] | undefined;
        let yearsDropdown: CalendarDropdownOption[] | undefined;

        if (showDropdowns) {
            [monthsDropdown, yearsDropdown] = this.getDropdowns(this.props);
        }

        return render({
            days: this.getDaysofMonth(this.props),
            month: startOfMonth(month),
            monthsDropdown,
            yearsDropdown
        });
    }
}
