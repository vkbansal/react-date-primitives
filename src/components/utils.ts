import { DayOfMonthSubset } from './types';

const monthsCache: Map<string, DayOfMonthSubset[][]> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDatesofMonth(month: Date): DayOfMonthSubset[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DayOfMonthSubset[][] = Array.from(
        { length: 6 },
        (_1, i): DayOfMonthSubset[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonthSubset => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
