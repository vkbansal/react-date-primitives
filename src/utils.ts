export interface DayOfMonth {
    date: Date;
    inCurrentMonth: boolean;
}

export type DaysOfMonth = DayOfMonth[][];

const monthsCache: Map<string, DaysOfMonth> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDaysOfMonth(month: Date): DayOfMonth[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DaysOfMonth = Array.from(
        { length: 6 },
        (_1, i): DayOfMonth[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonth => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
