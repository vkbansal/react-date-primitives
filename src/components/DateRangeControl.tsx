import * as React from 'react';

import { startOfMonth, addMonths, isDayBefore, callIfExists } from '../utils';
import { DateRangeControlProps, DateRangeControlState } from './types';

/**
 * Controller component for making a date-range picker in conjunction with [CalendarMonth]()
 */
export class DateRangeControl extends React.Component<
    DateRangeControlProps,
    DateRangeControlState
> {
    private numOfMonths: 2 | 3;

    constructor(props: DateRangeControlProps) {
        super(props);

        const { startDate, endDate, showThreeMonths } = props;

        this.numOfMonths = showThreeMonths ? 3 : 2;
        const thisMonth = startOfMonth(new Date());

        this.state = {
            startDate,
            endDate,
            months: Array.from({ length: this.numOfMonths }, (_, i) =>
                addMonths(startDate || thisMonth, i)
            ),
            selectionActive: false
        };
    }

    componentWillReceiveProps(nextProps: DateRangeControlProps) {
        const { startDate, endDate } = nextProps;

        this.setState((state) => {
            return {
                ...state,
                startDate,
                endDate,
                months: Array.from({ length: this.numOfMonths }, (_, i) =>
                    addMonths(startDate || state.months[0], i)
                )
            };
        });
    }

    handleDayClick = (day: Date) => {
        this.setState((state) => {
            const { startDate, selectionActive } = state;

            let newState: Pick<DateRangeControlProps, 'startDate' | 'endDate'>;

            if (startDate && selectionActive && !isDayBefore(day, startDate)) {
                newState = {
                    startDate,
                    endDate: day
                };

                callIfExists(this.props.onDatesChange, newState);

                return {
                    selectionActive: false,
                    ...newState
                };
            }

            newState = {
                startDate: day,
                endDate: undefined
            };

            callIfExists(this.props.onDatesChange, newState);

            return {
                selectionActive: true,
                ...newState
            };
        });
    };

    handleDayHover = (day: Date) => {
        const { selectionActive, startDate } = this.state;

        if (selectionActive && startDate && !isDayBefore(day, startDate)) {
            this.setState(() => ({
                endDate: day
            }));
        }
    };

    handleNavigation = (months: number) => () => {
        this.setState((state) => {
            return {
                months: state.months.map((month) => addMonths(month, months))
            };
        });
    };

    render() {
        const { render, minDate, maxDate } = this.props;
        const { months, startDate, endDate } = this.state;

        return render({
            months,
            startDate,
            endDate,
            minDate,
            maxDate,
            onDayClick: this.handleDayClick,
            onDayHover: this.handleDayHover,
            moveForward: this.handleNavigation(1),
            moveBackward: this.handleNavigation(-1)
        });
    }
}
