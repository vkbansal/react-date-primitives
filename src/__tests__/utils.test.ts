import * as utils from '../components/utils';

describe('utils test', () => {
    const month = new Date(Date.UTC(2017, 0 /* Jan */, 1, 0, 0, 0, 0));

    afterEach(() => {
        expect(month).toBe(month); // immutability check
    });

    test('setDay', () => {
        const newDate = utils.setDay(month, 10);
        expect(newDate.getDate()).toBe(10);
        expect(newDate.getMonth()).toBe(month.getMonth());
        expect(newDate.getFullYear()).toBe(month.getFullYear());
    });

    test('addDays', () => {
        let newDate = utils.setDay(month, 10);
        newDate = utils.addDays(newDate, 10);
        expect(newDate.getDate()).toBe(20);
        expect(newDate.getMonth()).toBe(month.getMonth());
        expect(newDate.getFullYear()).toBe(month.getFullYear());
    });

    test('setMonth', () => {
        const newDate = utils.setMonth(month, 10);
        expect(newDate.getMonth()).toBe(10);
        expect(newDate.getDate()).toBe(month.getDate());
        expect(newDate.getFullYear()).toBe(month.getFullYear());
    });

    test('addMonths', () => {
        let newDate = utils.setMonth(month, 2);
        newDate = utils.addMonths(newDate, 3);
        expect(newDate.getMonth()).toBe(5);
        expect(newDate.getDate()).toBe(month.getDate());
        expect(newDate.getFullYear()).toBe(month.getFullYear());
    });

    test('setYear', () => {
        const newDate = utils.setYear(month, 2206);
        expect(newDate.getFullYear()).toBe(2206);
        expect(newDate.getDate()).toBe(month.getDate());
        expect(newDate.getMonth()).toBe(month.getMonth());
    });

    test('startOfMonth', () => {
        const date = new Date();
        const newDate = utils.startOfMonth(date);
        expect(newDate.getDate()).toBe(1);
        expect(newDate.getMonth()).toBe(date.getMonth());
        expect(newDate.getFullYear()).toBe(date.getFullYear());
    });

    test('endOfMonth', () => {
        const newDate = utils.endOfMonth(month);
        expect(newDate.getDate()).toBe(31);
        expect(newDate.getMonth()).toBe(month.getMonth());
        expect(newDate.getFullYear()).toBe(month.getFullYear());
    });

    test('isSameMonth', () => {
        expect(
            utils.isSameMonth(new Date(2014, 8 /* Sep */, 2), new Date(2014, 8 /* Sep */, 25))
        ).toBe(true);

        expect(
            utils.isSameMonth(new Date(2014, 8 /* Sep */, 2), new Date(2013, 8 /* Sep */, 25))
        ).toBe(false);
    });

    test('isSameDay', () => {
        expect(
            utils.isSameDay(
                new Date(2014, 8 /* Sep */, 4, 6, 0),
                new Date(2014, 8 /* Sep */, 4, 18, 0)
            )
        ).toBe(true);

        expect(
            utils.isSameDay(
                new Date(2014, 8 /* Sep */, 4, 23, 59),
                new Date(2014, 8 /* Sep */, 5, 0, 0)
            )
        ).toBe(false);
    });

    test('isDayAfter', () => {
        expect(
            utils.isDayAfter(new Date(1989, 6 /* Jul */, 10), new Date(1987, 1 /* Feb */, 11))
        ).toBe(true);

        expect(
            utils.isDayAfter(new Date(1987, 1 /* Feb */, 11), new Date(1989, 6 /* Jul */, 10))
        ).toBe(false);

        expect(
            utils.isDayAfter(new Date(1989, 6 /* Jul */, 10), new Date(1989, 6 /* Jul */, 10))
        ).toBe(false);
    });

    test('isDayBefore', () => {
        expect(
            utils.isDayBefore(new Date(1987, 1 /* Feb */, 11), new Date(1989, 6 /* Jul */, 10))
        ).toBe(true);

        expect(
            utils.isDayBefore(new Date(1989, 6 /* Jul */, 10), new Date(1987, 1 /* Feb */, 11))
        ).toBe(false);

        expect(
            utils.isDayBefore(new Date(1989, 6 /* Jul */, 10), new Date(1989, 6 /* Jul */, 10))
        ).toBe(false);
    });

    test('callIfExists', () => {
        const callback = jest.fn();
        const args = [1, 2, 5, 'h'];

        expect(() => utils.callIfExists(undefined)).not.toThrow();
        utils.callIfExists(callback, ...args);
        expect(callback).toHaveBeenCalledWith(...args);
    });

    test('getDatesofMonth', () => {
        expect(utils.getDatesofMonth(month)).toMatchSnapshot();
    });
});
