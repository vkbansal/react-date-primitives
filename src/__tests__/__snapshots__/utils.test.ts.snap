// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils test getDaysOfMonth weeks start with FRIDAY 1`] = `

|     FRIDAY |   SATURDAY |     SUNDAY |     MONDAY |    TUESDAY |  WEDNESDAY |   THURSDAY |
| 2020-01-03 | 2020-01-04 | 2020-01-05 | 2020-01-06 | 2020-01-07 | 2020-01-08 | 2020-01-09 |
| 2020-01-10 | 2020-01-11 | 2020-01-12 | 2020-01-13 | 2020-01-14 | 2020-01-15 | 2020-01-16 |
| 2020-01-17 | 2020-01-18 | 2020-01-19 | 2020-01-20 | 2020-01-21 | 2020-01-22 | 2020-01-23 |
| 2020-01-24 | 2020-01-25 | 2020-01-26 | 2020-01-27 | 2020-01-28 | 2020-01-29 | 2020-01-30 |
| 2020-01-31 | 2020-02-01 | 2020-02-02 | 2020-02-03 | 2020-02-04 | 2020-02-05 | 2020-02-06 |
| 2020-02-07 | 2020-02-08 | 2020-02-09 | 2020-02-10 | 2020-02-11 | 2020-02-12 | 2020-02-13 |
`;

exports[`utils test getDaysOfMonth weeks start with MONDAY 1`] = `

|     MONDAY |    TUESDAY |  WEDNESDAY |   THURSDAY |     FRIDAY |   SATURDAY |     SUNDAY |
| 2019-12-30 | 2019-12-31 | 2020-01-01 | 2020-01-02 | 2020-01-03 | 2020-01-04 | 2020-01-05 |
| 2020-01-06 | 2020-01-07 | 2020-01-08 | 2020-01-09 | 2020-01-10 | 2020-01-11 | 2020-01-12 |
| 2020-01-13 | 2020-01-14 | 2020-01-15 | 2020-01-16 | 2020-01-17 | 2020-01-18 | 2020-01-19 |
| 2020-01-20 | 2020-01-21 | 2020-01-22 | 2020-01-23 | 2020-01-24 | 2020-01-25 | 2020-01-26 |
| 2020-01-27 | 2020-01-28 | 2020-01-29 | 2020-01-30 | 2020-01-31 | 2020-02-01 | 2020-02-02 |
| 2020-02-03 | 2020-02-04 | 2020-02-05 | 2020-02-06 | 2020-02-07 | 2020-02-08 | 2020-02-09 |
`;

exports[`utils test getDaysOfMonth weeks start with SATURDAY 1`] = `

|   SATURDAY |     SUNDAY |     MONDAY |    TUESDAY |  WEDNESDAY |   THURSDAY |     FRIDAY |
| 2020-01-04 | 2020-01-05 | 2020-01-06 | 2020-01-07 | 2020-01-08 | 2020-01-09 | 2020-01-10 |
| 2020-01-11 | 2020-01-12 | 2020-01-13 | 2020-01-14 | 2020-01-15 | 2020-01-16 | 2020-01-17 |
| 2020-01-18 | 2020-01-19 | 2020-01-20 | 2020-01-21 | 2020-01-22 | 2020-01-23 | 2020-01-24 |
| 2020-01-25 | 2020-01-26 | 2020-01-27 | 2020-01-28 | 2020-01-29 | 2020-01-30 | 2020-01-31 |
| 2020-02-01 | 2020-02-02 | 2020-02-03 | 2020-02-04 | 2020-02-05 | 2020-02-06 | 2020-02-07 |
| 2020-02-08 | 2020-02-09 | 2020-02-10 | 2020-02-11 | 2020-02-12 | 2020-02-13 | 2020-02-14 |
`;

exports[`utils test getDaysOfMonth weeks start with SUNDAY 1`] = `

|     SUNDAY |     MONDAY |    TUESDAY |  WEDNESDAY |   THURSDAY |     FRIDAY |   SATURDAY |
| 2019-12-29 | 2019-12-30 | 2019-12-31 | 2020-01-01 | 2020-01-02 | 2020-01-03 | 2020-01-04 |
| 2020-01-05 | 2020-01-06 | 2020-01-07 | 2020-01-08 | 2020-01-09 | 2020-01-10 | 2020-01-11 |
| 2020-01-12 | 2020-01-13 | 2020-01-14 | 2020-01-15 | 2020-01-16 | 2020-01-17 | 2020-01-18 |
| 2020-01-19 | 2020-01-20 | 2020-01-21 | 2020-01-22 | 2020-01-23 | 2020-01-24 | 2020-01-25 |
| 2020-01-26 | 2020-01-27 | 2020-01-28 | 2020-01-29 | 2020-01-30 | 2020-01-31 | 2020-02-01 |
| 2020-02-02 | 2020-02-03 | 2020-02-04 | 2020-02-05 | 2020-02-06 | 2020-02-07 | 2020-02-08 |
`;

exports[`utils test getDaysOfMonth weeks start with SUNDAY by default 1`] = `

|     SUNDAY |     MONDAY |    TUESDAY |  WEDNESDAY |   THURSDAY |     FRIDAY |   SATURDAY |
| 2019-12-29 | 2019-12-30 | 2019-12-31 | 2020-01-01 | 2020-01-02 | 2020-01-03 | 2020-01-04 |
| 2020-01-05 | 2020-01-06 | 2020-01-07 | 2020-01-08 | 2020-01-09 | 2020-01-10 | 2020-01-11 |
| 2020-01-12 | 2020-01-13 | 2020-01-14 | 2020-01-15 | 2020-01-16 | 2020-01-17 | 2020-01-18 |
| 2020-01-19 | 2020-01-20 | 2020-01-21 | 2020-01-22 | 2020-01-23 | 2020-01-24 | 2020-01-25 |
| 2020-01-26 | 2020-01-27 | 2020-01-28 | 2020-01-29 | 2020-01-30 | 2020-01-31 | 2020-02-01 |
| 2020-02-02 | 2020-02-03 | 2020-02-04 | 2020-02-05 | 2020-02-06 | 2020-02-07 | 2020-02-08 |
`;

exports[`utils test getDaysOfMonth weeks start with THURSDAY 1`] = `

|   THURSDAY |     FRIDAY |   SATURDAY |     SUNDAY |     MONDAY |    TUESDAY |  WEDNESDAY |
| 2020-01-02 | 2020-01-03 | 2020-01-04 | 2020-01-05 | 2020-01-06 | 2020-01-07 | 2020-01-08 |
| 2020-01-09 | 2020-01-10 | 2020-01-11 | 2020-01-12 | 2020-01-13 | 2020-01-14 | 2020-01-15 |
| 2020-01-16 | 2020-01-17 | 2020-01-18 | 2020-01-19 | 2020-01-20 | 2020-01-21 | 2020-01-22 |
| 2020-01-23 | 2020-01-24 | 2020-01-25 | 2020-01-26 | 2020-01-27 | 2020-01-28 | 2020-01-29 |
| 2020-01-30 | 2020-01-31 | 2020-02-01 | 2020-02-02 | 2020-02-03 | 2020-02-04 | 2020-02-05 |
| 2020-02-06 | 2020-02-07 | 2020-02-08 | 2020-02-09 | 2020-02-10 | 2020-02-11 | 2020-02-12 |
`;

exports[`utils test getDaysOfMonth weeks start with TUESDAY 1`] = `

|    TUESDAY |  WEDNESDAY |   THURSDAY |     FRIDAY |   SATURDAY |     SUNDAY |     MONDAY |
| 2019-12-31 | 2020-01-01 | 2020-01-02 | 2020-01-03 | 2020-01-04 | 2020-01-05 | 2020-01-06 |
| 2020-01-07 | 2020-01-08 | 2020-01-09 | 2020-01-10 | 2020-01-11 | 2020-01-12 | 2020-01-13 |
| 2020-01-14 | 2020-01-15 | 2020-01-16 | 2020-01-17 | 2020-01-18 | 2020-01-19 | 2020-01-20 |
| 2020-01-21 | 2020-01-22 | 2020-01-23 | 2020-01-24 | 2020-01-25 | 2020-01-26 | 2020-01-27 |
| 2020-01-28 | 2020-01-29 | 2020-01-30 | 2020-01-31 | 2020-02-01 | 2020-02-02 | 2020-02-03 |
| 2020-02-04 | 2020-02-05 | 2020-02-06 | 2020-02-07 | 2020-02-08 | 2020-02-09 | 2020-02-10 |
`;

exports[`utils test getDaysOfMonth weeks start with WEDNESDAY 1`] = `

|  WEDNESDAY |   THURSDAY |     FRIDAY |   SATURDAY |     SUNDAY |     MONDAY |    TUESDAY |
| 2020-01-01 | 2020-01-02 | 2020-01-03 | 2020-01-04 | 2020-01-05 | 2020-01-06 | 2020-01-07 |
| 2020-01-08 | 2020-01-09 | 2020-01-10 | 2020-01-11 | 2020-01-12 | 2020-01-13 | 2020-01-14 |
| 2020-01-15 | 2020-01-16 | 2020-01-17 | 2020-01-18 | 2020-01-19 | 2020-01-20 | 2020-01-21 |
| 2020-01-22 | 2020-01-23 | 2020-01-24 | 2020-01-25 | 2020-01-26 | 2020-01-27 | 2020-01-28 |
| 2020-01-29 | 2020-01-30 | 2020-01-31 | 2020-02-01 | 2020-02-02 | 2020-02-03 | 2020-02-04 |
| 2020-02-05 | 2020-02-06 | 2020-02-07 | 2020-02-08 | 2020-02-09 | 2020-02-10 | 2020-02-11 |
`;

exports[`utils test getDaysOfRangeMonth works without startDate and endDate 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
Apr =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |    
May =>     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |    
Jun =>     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |     |     |    
Jul =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Aug =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |    
Sep =>     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |     |    
Oct =>     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |    
Nov =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |     |     |     |    
Dec =>     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |    
`;
